core/src/main/java/hudson/util/DirScanner.java

```diff
@@ -130,6 +139,53 @@ public void scan(File dir, FileVisitor visitor) throws IOException {
 
         private static final long serialVersionUID = 1L;
     }
+    
+    private static class DescendantFileSelector implements FileSelector{
+        private final Set<String> alreadyDeselected;
+        private final FilePath baseDirFP;
+        private final int baseDirPathLength;
+
+        private DescendantFileSelector(File basedir){
+            this.baseDirFP = new FilePath(basedir);
+            this.baseDirPathLength = basedir.getPath().length();
+            this.alreadyDeselected = new HashSet<>();
+        }
+        
+        @Override
+        public boolean isSelected(File basedir, String filename, File file) throws BuildException {
+            String parentName = file.getParent();
+            if (parentName.length() > baseDirPathLength) {
+                // remove the trailing slash
+                String parentRelativeName = parentName.substring(baseDirPathLength + 1);
+
+                // as the visit is done following depth-first approach, we just have to check the parent once
+                // and then simply using the set
+                // in case something went wrong with the order, the isDescendant is called with just a lost
+                // in terms of performance, no impact on the result
+                if (alreadyDeselected.contains(parentRelativeName)) {
+                    alreadyDeselected.add(filename);
+                    return false;
+                }
+            }
+            // else: we have the direct children of the basedir
+
+            if (file.isDirectory()) {
+                try {
+                    if (baseDirFP.isDescendant(filename)) {
+                        return true;
+                    } else {
+                        alreadyDeselected.add(filename);
+                        return false;
+                    }
+                }
+                catch (IOException | InterruptedException e) {
```

Comment:
Should not be interrupted

Reply:
@oleg-nenashev What do you propose here? Wrapping it inside a `RuntimeException`, throwing an `AssertException` ? The method signature does not allow the throw of InterruptedException here.